// @nestjs-modules/ioredis/dist/modules/redis-core.module.js
"use strict";
var RedisCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisCoreModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const redis_connection_provider_1 = require("../providers/redis-connection.provider");
const redis_options_provider_1 = require("../providers/redis-options.provider");
const redis_connection_util_1 = require("../utils/redis-connection.util");
let RedisCoreModule = RedisCoreModule_1 = class RedisCoreModule {
    constructor(moduleRef) {
        this.moduleRef = moduleRef;
    }
    async onApplicationShutdown() {
        for (const token of RedisCoreModule_1.connectionTokens) {
            try {
                const connection = this.moduleRef.get(token);
                if (connection && typeof connection.quit === 'function') {
                    await connection.quit();
                }
            }
            catch { }
        }
        RedisCoreModule_1.connectionTokens.clear();
    }
    static forRoot(options, connection) {
        const connectionToken = (0, redis_connection_util_1.getRedisConnectionToken)(connection);
        RedisCoreModule_1.connectionTokens.add(connectionToken);
        const optionsProvider = (0, redis_options_provider_1.createRedisOptionsProvider)(options, connection);
        const connectionProvider = (0, redis_connection_provider_1.createRedisConnectionProvider)(options, connection);
        return {
            module: RedisCoreModule_1,
            providers: [optionsProvider, connectionProvider],
            exports: [optionsProvider, connectionProvider],
        };
    }
    static forRootAsync(options, connection) {
        const connectionToken = (0, redis_connection_util_1.getRedisConnectionToken)(connection);
        RedisCoreModule_1.connectionTokens.add(connectionToken);
        const asyncConnectionProvider = (0, redis_connection_provider_1.createRedisAsyncConnectionProvider)(connection);
        return {
            module: RedisCoreModule_1,
            imports: options.imports,
            providers: [
                ...(0, redis_options_provider_1.createAsyncProviders)(options, connection),
                asyncConnectionProvider,
            ],
            exports: [asyncConnectionProvider],
        };
    }
};
exports.RedisCoreModule = RedisCoreModule;
RedisCoreModule.connectionTokens = new Set();
RedisCoreModule.createAsyncProviders = redis_options_provider_1.createAsyncProviders;
RedisCoreModule.createAsyncOptionsProvider = redis_options_provider_1.createAsyncOptionsProvider;
exports.RedisCoreModule = RedisCoreModule = RedisCoreModule_1 = tslib_1.__decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({}),
    tslib_1.__metadata("design:paramtypes", [core_1.ModuleRef])
], RedisCoreModule);
