// @nestjs-modules/ioredis/dist/providers/redis-options.provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRedisOptionsProvider = createRedisOptionsProvider;
exports.createAsyncProviders = createAsyncProviders;
exports.createAsyncOptionsProvider = createAsyncOptionsProvider;
const redis_connection_util_1 = require("../utils/redis-connection.util");
function createRedisOptionsProvider(options, connection) {
    return {
        provide: (0, redis_connection_util_1.getRedisOptionsToken)(connection),
        useValue: options,
    };
}
function createAsyncProviders(options, connection) {
    if (!(options.useExisting || options.useFactory || options.useClass)) {
        throw new Error('Invalid configuration. Must provide useFactory, useClass or useExisting');
    }
    if (options.useExisting || options.useFactory) {
        return [createAsyncOptionsProvider(options, connection)];
    }
    return [
        createAsyncOptionsProvider(options, connection),
        { provide: options.useClass, useClass: options.useClass },
    ];
}
function createAsyncOptionsProvider(options, connection) {
    if (!(options.useExisting || options.useFactory || options.useClass)) {
        throw new Error('Invalid configuration. Must provide useFactory, useClass or useExisting');
    }
    if (options.useFactory) {
        return {
            provide: (0, redis_connection_util_1.getRedisOptionsToken)(connection),
            useFactory: options.useFactory,
            inject: options.inject || [],
        };
    }
    return {
        provide: (0, redis_connection_util_1.getRedisOptionsToken)(connection),
        async useFactory(optionsFactory) {
            return await optionsFactory.createRedisModuleOptions();
        },
        inject: [options.useClass || options.useExisting],
    };
}
