// @nestjs-modules/ioredis/dist/testing/redis-test.module.js
"use strict";
var RedisTestModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisTestModule = void 0;
exports.createMockRedis = createMockRedis;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const redis_connection_util_1 = require("../utils/redis-connection.util");
const MOCK_REDIS_METHODS = [
    'get',
    'set',
    'del',
    'exists',
    'expire',
    'ttl',
    'keys',
    'hget',
    'hset',
    'hdel',
    'hgetall',
    'lpush',
    'rpush',
    'lpop',
    'rpop',
    'lrange',
    'sadd',
    'srem',
    'smembers',
    'sismember',
    'zadd',
    'zrem',
    'zrange',
    'zrangebyscore',
    'publish',
    'subscribe',
    'ping',
    'quit',
    'info',
    'flushdb',
    'flushall',
    'multi',
    'exec',
    'pipeline',
    'mget',
    'mset',
    'incr',
    'decr',
    'incrby',
    'decrby',
    'scan',
    'hscan',
    'sscan',
    'zscan',
];
function createMockRedis(overrides) {
    const mock = {};
    for (const method of MOCK_REDIS_METHODS) {
        mock[method] = jest.fn();
    }
    if (overrides) {
        Object.assign(mock, overrides);
    }
    return mock;
}
let RedisTestModule = RedisTestModule_1 = class RedisTestModule {
    static forTest(connection, overrides) {
        const token = (0, redis_connection_util_1.getRedisConnectionToken)(connection);
        const mockRedis = createMockRedis(overrides);
        return {
            module: RedisTestModule_1,
            global: true,
            providers: [
                {
                    provide: token,
                    useValue: mockRedis,
                },
            ],
            exports: [token],
        };
    }
};
exports.RedisTestModule = RedisTestModule;
exports.RedisTestModule = RedisTestModule = RedisTestModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({})
], RedisTestModule);
